apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx

```diff
@@ -28,30 +38,120 @@ class ProgressTableLevelBubble extends React.PureComponent {
     isPaired: PropTypes.bool,
     bubbleSize: PropTypes.oneOf(Object.values(BubbleSize)).isRequired,
     title: PropTypes.string,
-    url: PropTypes.string
+    url: PropTypes.string,
+    useCache: PropTypes.bool
   };
 
   static defaultProps = {
     bubbleSize: BubbleSize.full
   };
 
-  renderContent() {
+  isEnabled() {
+    return !this.props.isDisabled && !this.isLocked();
+  }
+
+  isLocked() {
+    return this.props.levelStatus === LevelStatus.locked;
+  }
+
+  getBubbleShape() {
+    return this.props.isUnplugged
+      ? BubbleShape.pill
+      : this.props.isConcept
+      ? BubbleShape.diamond
+      : BubbleShape.circle;
+  }
+
+  /**
+   * We can't use our usual `progressStyles.hoverStyle` for hover effect here
+   * because we can't use radium in our cached html, so instead we
+   * conditionally add a CSS class to accomplish the same thing.
+   */
+  getClassNames() {
+    return classNames('progress-bubble', {
+      enabled: this.isEnabled()
+    });
+  }
+
+  /**
+   * Determines the simplest key to use for this bubble configuration.
+   */
+  getCacheKey() {
     const {
+      levelId,
       levelStatus,
+      levelKind,
       isUnplugged,
       isBonus,
       isPaired,
       title,
       bubbleSize
     } = this.props;
+
+    let statusString = `status=${levelStatus}`;
+    if (levelKind === LevelKind.assessment) {
+      statusString = `assessment:${statusString}`;
+    }
+
+    if (bubbleSize === BubbleSize.letter) {
+      return `letter:title=${title}&${statusString}`;
+    } else if (isUnplugged) {
+      `unplugged:${statusString}`;
+    }
+
+    const shapeString = `shape=${this.getBubbleShape()}`;
+    const contentString = this.isLocked()
+      ? `locked:`
+      : isPaired
+      ? `paired:`
+      : isBonus
+      ? `bonus:`
+      : title
+      ? `title=${title}`
+      : `levelId=${levelId}`;
```

Comment:
Why do we need to include levelId in the cache key?  (Does levelId ever affect how the bubble looks?)

Reply:
at this point that is just a fallback. my original implementation was only using `levelId` and `status` for the key, because the only difference between the appearance of the same level would come from status (actually, we also need `isPaired`, which i'd missed in my original implementation). 

i didn't want key generation and lookup to become a performance hit when we're rendering thousands of these at once. but Maureen pointed out that that resulted in lots of duplicated bubbles in the cache (think if the number of not tried bubbles with a number in them repeated across lessons). switching from id + status to this implementation shrunk the average cache size in storybook from ~1500 to ~150.

i left `levelId` as a fallback since i'd already added it in all the places where this code is called, but that was lazy of me. if we still aren't using it once we settle on our cache key implementation, i will remove it before merging.

personally, i think keeping the keys as small as possible is worth the potential extra maintenance cost. _all_ of the slowness in this table comes from js scripting work, so imo every little bit of optimization here is worth it. but if you'd prefer a more future-proof implementation i can revisit and try to measure the performance impact.